File: java/Advent09.java
import java.util.*;
import java.util.stream.*;

import static java.util.Collections.max;
import static java.util.Collections.min;

public class Advent09 extends Advent {

  List<Long> nums = new ArrayList<>();

  public Advent09() {
    super(9);
  }

  @Override
  protected void parseInput() {
    nums = input.stream()
      .map(Long::valueOf)
      .collect(Collectors.toList());
  }

  @Override
  protected Object part1() {
    return IntStream.range(25, nums.size())
      .filter(this::valid)
      .mapToLong(nums::get)
      .findFirst()
      .getAsLong();
  }

  private boolean valid(int i) {
    var sublist = nums.subList(i-25,i);
    return sublist.stream()
      .flatMap(n -> sublist.stream().map(l -> l.longValue() + n))
      .noneMatch(nums.get(i)::equals);
  }

  @Override
  protected Object part2() {
    var target = (long) part1();
    int s = nums.size();
    for (int l = 2; l < s; l++) {
      for (int i = 0; i < s-l; i++) {
        var sublist = nums.subList(i, i+l);
        if (sublist.stream().mapToLong(Long::longValue).sum() == target) {
          return min(sublist)+max(sublist);
        }
      }
    }
    return null;
  }
}

File: 2020/python/advent_09.py
from advent_lib import *
import itertools as it

input = to_nums(lines(9))

def valid(n):
    prev = input[n-25:n]
    return input[n] not in [x+y for x in prev for y in prev]

part1 = [input[n] for n in range(25,len(input)) if valid(n)][0]
print("Part 1:", part1)

def part2():
    for l in range(2, len(input)):
        sublists = [input[i:i+l] for i in range(0,len(input)-l)]
        for sublist in sublists:
            if sum(sublist) == part1:
                return min(sublist)+max(sublist)


print("Part 2:", part2())
